Refactor `results-move` and bring to parity

portante opened the pull request:
We significantly refactor the implementation of the `pbench-results-move` command so that it matches the expected behaviors set by the legacy `pbench-move-results`.  This included ensure the tar ball was created with the correct directory structure, adding environmental metadata, and correcting the controller name appropriately.

We drop prompting for `--token`, we no longer require `--controller`, defaulting to the current host name.  We clean up the temporary directory, correct type hints, add doc strings, use existing agent logger, and emit status to stdout.

We also:
 * Move push to its own module
 * Rename `results_push` to `main`
 * Drop explicit `pbench-agent` logger from push
 * Refactor `move` results to own module
 * Provide `pbench-results-move` in the RPM
 * Add the new `--copy-only` switch
 * Rename `test_move_results.py` to `test_results_move.py`
 * Rename `--copy` to `--delete/--no-delete`
 * Update command line in help text of results cmds

webbnh commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -0,0 +1,196 @@
+                failures += 1
+                continue
+
+            try:
+                # We always remove the constructed tar ball, regardless of success
+                # or failure, since we keep the result directory below on failure.
+                os.remove(result_tb_name)
+            except OSError:
+                logger.error("Failed to remove {}", result_tb_name)
+                failures += 1
+                continue
```
I suspect that we should not count an exception at line 97 as a "failure" at line 100 -- after all, the result _did_ get copied....

webbnh replied:
Ditto lines 109 and 118.

portante replied:
Existing code ...

webbnh replied:
[Not really](https://github.com/distributed-system-analysis/pbench/pull/2325#discussion_r636139655) ...

At best, it's a _port_ of existing code.  The question remains, do we, going forward, want it to do what it is currently structured to do?

webbnh commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -0,0 +1,196 @@
+                try:
+                    copied.touch()
+                except OSError:
+                    logger.error("Failed to `touch` {}", copied)
```
BTW, why is creating markers in the host file system better than log messages or other indications of success -- does something check for these files later?  (E.g., should we be checking for them at line 44 and skipping the corresponding results if they've already been copied??)

portante replied:
This is existing behavior from `pbench-copy-results`.

webbnh replied:
Understood.  But, why do we have this behavior, and do we really want to propagate it to a new tool?

webbnh commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -0,0 +1,196 @@
+            else:
+                try:
+                    shutil.rmtree(result_dir)
+                except OSError:
+                    logger.error(
+                        "Failed to remove the {} directory hierarchy", result_dir
+                    )
```
This message is not reporting _why_ it failed (i.e., it should be attaching a stringized rendition of the exception object).

portante replied:
Existing code.

webbnh replied:
Needs improvement.

:-)

webbnh commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -0,0 +1,196 @@
+    prompt=False,
+    help="Controller name",
+)
+@click.option(
+    "--user", prompt=False, help="Pbench user (Obsolete)",
+)
```
If this is a new tool, there should be no obsolete switches.

webbnh replied:
Ditto for `--prefix`, `--xz-single-threaded`, and perhaps `--show-server`.

portante replied:
Existing code; lift-n-shift.

webbnh commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -0,0 +1,196 @@
+@click.option(
+    "--user", prompt=False, help="Pbench user (Obsolete)",
+)
+@click.option(
+    "--token",
+    required=True,
+    prompt=True,
+    envvar="PBENCH_ACCESS_TOKEN",
+    help="pbench server authentication token (will prompt if unspecified)",
+)
```
If we're going to prompt for the token, should we also prompt for the controller?

portante replied:
I am not a fan of prompting.

Existing code.

webbnh commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -0,0 +1,196 @@
+    xz_single_threaded: str,
+    show_server: str,
+):
+    """
+        TODO:  This is latent code -- it is currently unused and largely
+            untested, intended to implement a future tool to replace
+            pbench-make-result-tb and/or pbench-move-results.
+    """
```
With this PR, this will not longer be latent code.... Also, this comment block will be printed in response to the `--help` switch, so it should be formatted appropriately (see `main()` in `lib/pbench/cli/agent/commands/results/push.py`).

webbnh commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -0,0 +1,196 @@
+)
+@pass_cli_context
+def main(
+    context: click.Context,
```
Until something changes, this type should (I think...) be `CliContext` (from `pbench.cli.agent`).

webbnh commented on lib/pbench/cli/agent/commands/results/push.py:
```diff
@@ -0,0 +1,61 @@
+)
+@pass_cli_context
+def main(
+    context: click.Context, controller: str, result_tb_name: str, token: str,
```
Until something changes, the type of `context` should (I think...) be `CliContext` (from `pbench.cli.agent`).

webbnh replied:
It didn't look like you executed this suggestion.

If you're not going to change the type hint for `context`, then I suggest that you just remove it -- having it be wrong is likely to be worse than omitting it.

webbnh replied:
(Line 11, too.)

webbnh commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -0,0 +1,196 @@
+            untested, intended to implement a future tool to replace
+            pbench-make-result-tb and/or pbench-move-results.
+    """
+    real_context = click.get_current_context()
```
The context returned by `click.get_current_context()` is no more or less "real" than the one passed in as the first parameter....  :-D

Since we're only looking for a single item from the Click context (and, hopefully, our need for that will evaporate), I suggest just pulling it from `click.get_current_context().info_name` and dropping the intermediary.

webbnh reviewed (commented):
The previous approach of having all three tools in the same module was OK.  Splitting off `push` makes some sense, but having `copy` in a separate module from `move` with the common code in the `move` module seems...asymmetrical.

There are a bunch of comments below, but many of them come in duplicate or triplicate because they reference code which occurs in both `copy` and `move` or in all three tools.

I don't see unit tests for `copy`.

dbutenhof commented on lib/pbench/cli/agent/commands/results/copy.py:
```diff
@@ -0,0 +1,73 @@
+    prompt=False,
+    help="Controller name",
+)
+@click.option(
+    "--user", prompt=False, help="Pbench user (Obsolete)",
+)
```
It's an interesting philosophical question; yes, `pbench-results-copy` and `pbench-results-move` are now tools, not easily confused with the old `pbench-move-results` and `pbench-copy-results`. Does that mean it's pointless to imply any compatibility? If someone just changes `pbench-move-results --user me` to `pbench-results-move --user me` do (or should) we care that the "obsolete" option will generate a usage message instead of being diagnosed after parsing as you're doing now?

We're not really in the position of a staged decprecation here, since we're both renaming and radically changing the behavior of the tools.

So, yeah, it can be argued that we should just drop these entirely and let them be parser errors...

webbnh replied:
I would suggest that, if someone is in the position of changing `pbench-move-results` to `pbench-results-move`, then they are also in a position where they can remove the `--user me` option (and, of course, they are going to have to supply a token somehow, as well...).

dbutenhof commented on lib/pbench/cli/agent/commands/results/copy.py:
```diff
@@ -0,0 +1,73 @@
+@click.option(
+    "--user", prompt=False, help="Pbench user (Obsolete)",
+)
+@click.option(
+    "--token",
+    required=True,
+    prompt=True,
+    envvar="PBENCH_ACCESS_TOKEN",
+    help="pbench server authentication token (will prompt if unspecified)",
+)
```
Makes sense: `controller` can be defaulted from the current controller hostname, but there's no default for `token`.

portante replied:
Removed all the prompting for `pbench-results-move`.

dbutenhof reviewed (commented):
As usual, Webb has good comments; in particular, though, I think that I'd have approved if you hadn't left the stale docstrings and comments.

portante commented:
> The previous approach of having all three tools in the same module was OK. Splitting off `push` makes some sense, but having `copy` in a separate module from `move` with the common code in the `move` module seems...asymmetrical.

The `copy` / `move` thing aside, if you consider how all the other click based commands are handled, they all are using `...<module>:main`.  So this is just a change to be consistent.

> There are a bunch of comments below, but many of them come in duplicate or triplicate because they reference code which occurs in both `copy` and `move` or in all three tools.

I'll be proposing an update which drops the separate copy command (I have never been a fan of two command names invoking the same code).

> I don't see unit tests for `copy`.

Yes, this is a `draft` PR, and was posted in order for us to see the possibility of what could be done, and build something to see if it works.

Before this PR is moved out of `draft` mode, the proper unit tests the `copy` operation will be made.

portante commented on agent/Makefile:
```diff
@@ -66,6 +66,8 @@ click-scripts = \
 	pbench-config \
 	pbench-list-tools \
 	pbench-list-triggers \
+	pbench-results-copy \
```
Dropping this in favor of a `--copy` switch.

portante commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -0,0 +1,196 @@
+
+            try:
+                mrt = MakeResultTb(result_dir, temp_dir, self.config, logger)
+            except FileNotFoundError as e:
+                logger.error("File Not Found Error, {}", e)
+                continue
+            except NotADirectoryError as e:
+                logger.error("Bad Directory, {}", e)
+                continue
```
Keep in mind, this is existing code.  I did not make any changes to this code.

I provided multiple commits to show that it was a lift and shift.

webbnh replied:
Understood.  But, this **is**, in effect, _new code_.  The source of the "lift and shift" was latent code which was waiting for its initial deployment; it should be subject to inspection, shakedown, and revision before we put it into use.

portante commented on lib/pbench/test/unit/agent/task/test_move_results.py:
```diff
@@ -6,7 +6,7 @@
 from click.testing import CliRunner
 
-from pbench.cli.agent.commands import results
+from pbench.cli.agent.commands.results import move
```
Seems fine to me.  That is how it was done previously, so this change was trying to be minimal and not rototill the approach.

webbnh replied:
IIRC, this was "latent code" which means that it doesn't hold much power of precedent.

We normally use `from` to import functions or classes (or global variables); here we are using it to import a _module_...why not use a conventional bare `import` for that?

portante replied:
> IIRC, this was "latent code" which means that it doesn't hold much power of precedent.

Aren't the other modules that use the same approach considered the precedent?

Is this change really required to make this PR acceptable?  The mechanism in place does not appear to change how the code operates or its correctness.  It seems just to be mechanism that is important to be consistent about going forward for code maintainability, would you agree?  As such can we agree to not introduce a new way to do things for all other commands in this PR?  Wouldn't that kind of change be done as a separate PR addressing all the commands?

webbnh replied:
I'm mostly concerned with determining "best practice" and encouraging its use.  I'm not opposed to having separate PRs which make existing code conform to our expectations.  But, that doesn't mean that we should accept new or changed code which does not conform (the fact that we intend to fix them later doesn't mean that it's OK to merge in departures now).

That said, my understanding of the syntax in use here (and, I grant, the change you made does follow the previous usage -- it's not a change which makes things "worse" -- but I don't think it's the form that we want) is that it requests the import of a module from a package.  Part of my objection is that I don't think that target of the `from` actually is a package in the typical Python sense.  Nevertheless, what we're interested in here is importing an entire, specific module (and not a subset of something else), and therefore we should be using a bare `import` and not the `from` form.

What is your concern about switching away from `from`, here?

portante replied:
I don't want to slowly skew the code "best practices".  I'd rather have all the code behaving consistently, and then, when we decide that we want to change the usage pattern to an agreed upon best practice, move the code that way together.

webbnh replied:
If it were a question of preserving existing practice vs. a gradual slew, I'd be willing to be persuaded, but there's not a lot of consistency currently.

I found only about a dozen cases where we are importing _modules_ using the `from` form; I found three cases 
where we use the bare `import` form; and, in the vast majority of cases (>80%), we use the `from` form targeting definitions _within_ a module.

So, I guess "our practice" is to use the `from` form to import the specific definitions that we need, which doesn't sound bad, but, in the case of test programs, we should probably be importing the whole module (because we should be testing more or less everything in it).  If we're going to import the whole module, I'd recommend importing it using a bare `import`, because I think that that syntax better expresses that intention.

portante replied:
I've re-read this thread a number of times, and I think I understand now what you are trying to say.  I leaped to a different conclusion from the discussion which was not correct.

Let's what folks think about the `from` import of the leaf module.

webbnh replied:
My initial response to that is, why did we name these functions `main()`, when doing so is bound to cause problems like this?  😉 

I think the Pythonic response is to use the `as` clause:  `from foo.bar.baz.move import main as move_baz`.

portante commented on setup.cfg:
```diff
@@ -31,7 +31,9 @@ console_scripts =
    pbench-list-tools = pbench.cli.agent.commands.tools.list:main
    pbench-list-triggers = pbench.cli.agent.commands.triggers.list:main
    pbench-register-tool-trigger = pbench.cli.agent.commands.triggers.register:main
-   pbench-results-push = pbench.cli.agent.commands.results:results_push
```
Notice that `pbench-results-push` was the aberration before.

webbnh replied:
See [my comment](https://github.com/distributed-system-analysis/pbench/pull/2325#issuecomment-845149633) below.

webbnh commented:
> The copy / move thing aside, if you consider how all the other click based commands are handled, they all are using `...<module>:main`. So this is just a change to be consistent.

I'm concerned that it's needless consistency.  The Click interplay with `setup` offers the flexibility to specify any Python function -- it doesn't have to be named `main()` or be in any particular position in a module.  In fact, naming the target functions `main()` engenders confusion that the modules might or should be invocable as "scripts" (i.e., standalone commands) when they are not -- the real main is created during the installation.  If you feel the need for some sort of consistency or convention, that's fine, but I don't think that using `main()` is a good one.

> I'll be proposing an update which drops the separate copy command (I have never been a fan of two command names invoking the same code).

Okie.  I would say that shared code is generally good, but how it's structured does affect the reader's perceptions.  My objection is to "control parameters".  However, in this case, I've been quiet about that, because the two operations do almost exactly the same thing except for one detail buried inside an inner loop, so it's likely that using a control parameter is the "least of evils".  "Better" alternatives might be decomposing the function into parts (e.g., a file iterator, a principal operation encapsulated in a subroutine, etc.) which would be used by each of the move and copy operations, or passing a call-back instead of the control parameter to common code, but each of those is more complicated and potentially less maintainable (although, Python with its first-class functions, passing an inner function might actually be better than using a control parameter and having to use a `if` statement).

> this is a draft PR, and was posted in order for us to see the possibility of what could be done, and build something to see if it works.

Yeah...we just wanted to offer some input on direction, before things got too calcified.  :-)  (And, re unit tests, I just wanted to make sure the item was on your agenda -- I would hate for this to get to final review and _then_ have us go, "Oh, yeah...."

portante reviewed (commented):
Okay, I believe this code is no longer a draft, ready for full review.

dbutenhof commented on lib/pbench/agent/results.py:
```diff
@@ -14,45 +14,79 @@
+            result_dir -- directory where tb file is collected.
+            target_dir -- directory where tar file needs to be moved.
+            config -- PbenchAgent config object
+            logger -- logger objects helps logging important details
```
Not sure why this is showing as a diff, since these lines didn't change; but if you pass this way again, you might fix the grammar of the "logger" description.

dbutenhof commented on lib/pbench/agent/results.py:
```diff
@@ -72,23 +106,28 @@ def check_result_target_dir(dir_path: str, name: str) -> Path:
-        if os.path.exists(f"{self.result_dir}.copied"):
-            raise Exception(f"Already copied {str(self.result_dir)}")
+        The metadata.log file in the result directory is double checked to be
+        sure it is properly setup, and then add the "run.raw_size" and the
```
"valid" might be better than "properly set[]up"

dbutenhof commented on lib/pbench/agent/results.py:
```diff
@@ -72,23 +106,28 @@ def check_result_target_dir(dir_path: str, name: str) -> Path:
 
-    def make_result_tb(self) -> str:
-        """make_result_tb - make the result tarball.
+    def make_result_tb(self, single_threaded: bool = False) -> str:
```
not "str"; maybe `-> Tuple[Path, int, str]`??

dbutenhof commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -0,0 +1,257 @@
+
+
+class MoveResults(BaseCommand):
+    """MoveResutls - command implementation for "pbench results move."
```
"MoveResults" ... but I think we probably shouldn't be repeating the method name; just another "double truth" that can get out of sync later with the code.

dbutenhof commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -0,0 +1,257 @@
+                        "Error uploading file, '%s', '%s'", result_tb_name, exc
+                    )
+                    failures += 1
+                    # We don't know what this operation failed; regardless,
```
"why this operation failed"??? And it's repeated multiple times below. Also, the "regardless, given [...] this tar ball remotely," probably isn't really useful, and the longer comment is just harder to read. Maybe just "We don't know why the operation failed, but trying to copy another tarball isn't likely to succeed."

dbutenhof reviewed (approved):
Without weighing in on the import statement style (though I tend to agree with Webb ;-) ), this looks OK but there are some nits including a "dishonest" return type hint.

webbnh reviewed (changes requested):
Sorry for the volume below, but it seems like these changes could use some polishing.

Also, there are some weightier items:

- In general (and in the specific case cited below), do we want to document all exceptions raised by a given function?  This strikes me as a boundless problem with little payback.
- When a function has a potentially large number of possible exceptions which all get handled in the same way, how should we construct the caller in a way which is resilient?
- Should we be using the Python2 string format specifiers, or should we stick with the Python3 ones?
- Do we want to do all the detailed handling of setting up the subprocess invocations manually, or should we just construct the appropriate shell command line and invoke it in a subprocess running a shell (which seems much simpler and easier to understand and maintain)?
- Does running `tar | xz` do anything different from `tar --xz`?...is it worth supporting the two options?
- and a bunch of smaller stuff.

portante commented on lib/pbench/agent/results.py:
```diff
@@ -14,45 +14,79 @@
+            config -- PbenchAgent config object
+            logger -- logger objects helps logging important details
+
+        Raises:
+            AlreadyCopied       if we find a "*.copied" marker for the result
+                                directory.
+            BenchmarkRunning    if we find a "*/.running" marker in the result
+                                directory.
+            FileNotFoundError   if either the result or target directories do
+                                not exist
+            NotADirectoryError  if either the result or target directories are
+                                not actual directories
+            RuntimeError        if it cannot find the 'tar' command on the PATH
```
Not sure I follow.

Generic exception like `FileNotFoundError` and `NotADirectoryError` and `ValueError` are not self explanatory, and only when the exception happens will a string be provided to know what to do with it.

While working on this code, I found myself pouring over the existing to code find the behaviors on how to handle them in the "results move" loop, that it was easier to collect them once, then to keep reading the code as they were updated.

What is the harm in documenting them since they encode the behavior we are trying to implement?

portante commented on lib/pbench/agent/results.py:
```diff
@@ -108,9 +147,24 @@ def make_result_tb(self) -> str:
         result_size = sum(f.stat().st_size for f in self.result_dir.rglob("*"))
         self.logger.debug(
-            "Preparing to tar up {} bytes of data from {}",
+            "Preparing to tar up %d bytes of data from %s",
```
Unfortunately, the agent side logger does not support the `%s` style.  I ran into that in testing when they all blew up.

A separate commit will need to graduate the agent logger from `%`-style to `{}`-style.

portante commented on lib/pbench/agent/results.py:
```diff
@@ -122,45 +176,137 @@ def make_result_tb(self) -> str:
         with mdlog_name.open("w") as fp:
             mdlog.write(fp)
 
+        # FIXME: /bin/cp pbench.log ${pbench_run_name}/
+
```
Need to execute that still.

portante replied:
Actually, I am thinking this might be a good time to cut the cord on this behavior.

I have never liked the idea of copying this log file into the pbench result tar ball, without truncating it, and we can't do that because we allow the "results move" behaviors to happen while `bench-scripts` are executing.

I'll remove this comment in favor of tracking it via issue #1680.

portante commented on lib/pbench/agent/results.py:
```diff
@@ -122,45 +176,137 @@ def make_result_tb(self) -> str:
-            )
+            # Invoke tar directly for efficiency.
+            if single_threaded:
+                with tarball.open("w") as ofp, e_file.open("w") as efp:
+                    tar_proc = subprocess.Popen(
+                        args,
+                        cwd=str(self.result_dir.parent),
+                        stdin=None,
+                        stdout=ofp,
+                        stderr=efp,
+                    )
```
Python3 is not involved.  The two processes are connected by pipes and they execute.  The creating process does not get involved except to close at the end.

webbnh replied:
It looks to me like the Python interpreter is responsible for collecting the outputs and writing them to their respective files.  I have no idea how `tar` or `xz` would do those writes, but I would expect that they are highly optimized for doing bulk output; I have no idea how Python handles collecting output from subprocesses and writing it to files, but if I had to bet, I would bet on `tar` and/or `xz` over Python.

It seems to me that we're likely to get the best performance by letting `tar` do all of the I/O, which is why I suggested using `call()` to get Python out of the way.  I think the input comes from the file system, and the output goes to the file system, and all we care about is whether the command completes successfully, so I don't see how being involved in the plumbing (especially if it means that `stdout` is a "stream" instead of a block device) is going to help.

portante replied:
The pipes are connected together.  Python3 is not involved.

webbnh replied:
I'm willing to believe what you say (I'm having trouble wrapping my head completely around this at the moment), but I still think there is a difference.

I think the code as it stood is in effect running a command like `tar cf - > tarball.tar` which I _think_ produces a subtly different result from `tar cf tarball.tar`.  Now, it's quite possible that I'm way out in the weeds on this one, but in the second case, `tar` does the file open and is fully aware of the nature of the file it is writing to (including its device characteristics), while in the first case, the output is a shell redirect, and I'm not sure that `tar` knows anything about it other than that it's a byte stream.  If I'm right about this, then, in the second case, `tar` can use I/O optimizations which aren't available in the first case.

portante commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -0,0 +1,257 @@
+        context.controller = controller
+    else:
+        controller = os.environ.get("_pbench_full_hostname", socket.getfqdn())
+        if not controller:
+            click.echo(
+                "INTERNAL ERROR - unable to determine the controller name, could not fetch the FQDN of the host",
+                err=True,
+            )
```
We don't require the `--controller` switch, do we?  So wouldn't emitting a message telling the user to provide a value be against our interface goals?  If our fallback of calling `socket.getfqdn()` does not provide a FQDN name for the host, we don't have a fallback to use, which is not the fault of the user, right?  So why not inform the user of a work-around, as you suggest, while still letting them know a base assumption is broken?

webbnh replied:
To be honest, I get real fuzzy real fast on who is responsible for which domains with this stuff.

From what I've seen, it seems clear that we currently need the controller to be set correctly, otherwise we end up with problems trying to unpack the tarball (which, I have to admit, sort of surprises me, but I assume that's just my limited knowledge).  We have three spots to try to get the controller from -- the command line, the environment variable, and the current host -- I don't know who is responsible for each of them.  The command line seems like it would be the user, but if we're invoking this command _on behalf of the user_, then it's on us (although, I don't know where _we_ would get the value from).  The environment variable could be something that we set up, or it could be there for the user's convenience; I have no idea.  But, the configuration of the host is _not_ in our wheelhouse (right??), so, if the fallback doesn't work, that's not a short-coming our our product -- hence, _not_ an "internal error" -- that problem is something the user has to fix (I would think...) if they want us to get the controller name that way.

So, I'm all for telling the user what happened, and I'm in favor of explaining how to fix it, but labeling something an internal error when it's ultimately the result of something the user did (or didn't do), doesn't help the user, and it won't be good for us.

portante replied:
The user is not responsible for the host not having an FQDN, is it?

webbnh replied:
Well, _somebody_ configured the host that this stuff is running on.  I'm gonna go out on a limb here and say that, if it wasn't the user himself, at least he's got a closer relationship to the sysadmin than our code does.

We don't consider file system access problems to be internal errors (right?), so why should problems accessing other parts of the system configuration be deemed internal errors?

portante replied:
Arguably we should report them as INTERNAL ERRORS because they are not errors the user caused, but are fundamental broken assumptions.

It is okay to disagree on that point.

If just removing `INTERNAL ERROR` from the message is what you are looking for, then I'll remove it.

webbnh replied:
I would suggest that an internal error is the result of software detecting of an incorrect or inconsistent piece of state that it maintains.  However, the FSDN is an _external_ piece of state -- it's not ours to set or maintain.

We can quibble about whether the user caused it or not.  I'm not sure who "the user" is in this context...mostly it seems to me to be people who provision machines, install stuff on them, and then run Pbench...in which case, it would be up to them to make sure that the system has a FQDN configured.  In any case, it's the user (not Pbench or its maintainers) who has any realistic hope of getting the system configuration corrected if it's missing.

What I'm looking for is a good experience for the user, and, secondarily, for us.  If we flag an error in the user's configuration as an "internal Pbench error", then the user is unlikely to pursue fixing it themself and is instead (best case...) going to come after us to fix it.  (In the worst case, they will label our software as "buggy" and go elsewhere.)

I don't know how we expect to come up with the controller name, but, if we fail, and the user can address the problem, then I think we should empower and encourage the user to dig us out of the collective hole that we're in, and removing `INTERNAL ERROR` would be an excellent first step.

portante commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -0,0 +1,257 @@
+        clk_ctx.exit(1)
+    context.controller = controller
+
+    if not token:
+        click.echo(
+            "No token provided: use the --token option on the command line, or"
+            " define the PBENCH_ACCESS_TOKEN environment varible",
+            err=True,
+        )
+        clk_ctx.exit(1)
```
Ah, um, ..., this is a result of a bad environment variable setting on my part in my testing, which I later fixed, and never reverted this.

portante commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -0,0 +1,257 @@
+    "--copy",
+    is_flag=True,
+    help="Only perform the remote copy, do not remove local data",
+)
```
Seems reasonable.  Also added a test for `--no-delete`, and a test for the `.running` directory behavior.

portante commented on lib/pbench/test/unit/agent/task/test_make_result_tb.py:
```diff
@@ -1,74 +1,140 @@
             # Teardown the setup
-            self.config, self.logger = None, None
-            if os.path.exists(f"{os.path.realpath(MRT_DIR)}.copied"):
-                os.remove(f"{os.path.realpath(MRT_DIR)}.copied")
-            if os.path.exists(f"{os.path.realpath(MRT_DIR)}/.running"):
-                os.remove(f"{os.path.realpath(MRT_DIR)}/.running")
+            (
+                self.controller,
+                self.name,
+                self.result_dir,
+                self.target_dir,
+                self.config,
+                self.logger,
+            ) = (
+                None,
+                None,
+                None,
+                None,
+                None,
+                None,
+            )
```
Tidy.

webbnh replied:
Hmmm.  Doesn't seem like it's worth the space it takes up.  (And, its presence in the code gives the reader the impression that it's important/operable.)

portante commented on lib/pbench/test/unit/agent/task/test_move_results.py:
```diff
@@ -125,18 +108,32 @@ def test_move_results(monkeypatch, caplog, pytestconfig):
         )
-        assert result.exit_code == 0
-        assert result.stdout_bytes
+        assert (
+            result.exit_code == 0
+        ), f"Expected a successful operation, exit_code == {result.exit_code:d}"
```
Yes, I'll fix this.

BUT ...

This brings up another problem with the `CliRunner` ... it is completely hiding stack traces.

I had to resort to adding my own `traceback` module invocation to get them printed to `stdout` to debug problems.

I think this is because our default agent logging is to the `pbench.log` file, but not entirely sure.

portante replied:
Tracked in #2335.

portante commented on lib/pbench/agent/results.py:
```diff
@@ -14,45 +14,79 @@
         ), f"config, '{config!r}', and/or logger, '{logger!r}', not provided"
-        self.result_dir = self.check_result_target_dir(result_dir, "Result")
-        self.target_dir = self.check_result_target_dir(target_dir, "Target")
+        self.tar_path = find_executable("tar")
+        if self.tar_path is None:
+            raise RuntimeError(f"External 'tar' executable not found")
+        self.xz_path = find_executable("xz")
+        if self.xz_path is None:
+            raise RuntimeError(f"External 'xz' executable not found")
```
The problem is we don't want the defined a class definition time (when the module is imported), but once the "main" program has started running.

webbnh replied:
What do you think is going to change or happen between "class definition time" and when the "main" program starts running?

My understanding is that Python code is just a script which gets read from top to bottom executing each line in sequence -- things like `class` statements have effects when they are read -- and it's only when a class or function gets referenced (or invoked) that stuff "runs" as we think of it.  Import statements are just another thing which gets executed, as are initializers for global variables and class attributes.  So, yeah, moving these calls of `find_executable()` from instance initializers to class initializers would mean that they get executed at a different time, but that's exactly what I want -- they'll get executed only once overall instead of once for each instance.  Are you saying that the current working directory or the shell path (or something else which will affect their operation) will be different after the main function gets called?

portante replied:
> Are you saying that the current working directory or the shell path (or something else which will affect their operation) will be different after the main function gets called?

Sort of.

When `test_results_move` imports the `move` module, that intern imports the `MakeResultTb` class.  So just by loading the test, the `find_executable()` methods would be invoked at module load time.  This makes it hard to mock out the `tar` and `xz` commands.

In the case of the Tool Meister code, was have everything in one module, so we can do it once on when the command actually executes.

But here, we either have to pass in the command paths to use each time, or make the optional parameters to override the class definitions.  Hmm.

webbnh replied:
Ah, yes...I hadn't considered a requirement to be able to mock the `find_executable()` call.

Of course, if these attributes were class attributes, the test code could simply reassign them without having to patch or mock anything.

portante replied:
Unit tests can certainly solve that by slamming the class variables to what they require, but that prevents other kind of testing.

webbnh replied:
Outside of unit testing, I think having the values calculated only once (and stored as class attributes) is a better solution (because it avoids unnecessary time spent in instance initialization).  Now, this probably isn't performance-sensitive code, and, in most scenarios, we probably never create more than one instance of the class, and so we probably don't save much by doing the initialization during class initialization instead of during the (one) object instantiation, but that's not an argument _against_ using class initialization.

As for unit testing, being able to write the class attributes is probably exactly equivalent to monkeypatching class member functions with mocks, and it's simpler and probably clearer.

So, what other testing did you have in mind that using class attributes would interfere with?

webbnh replied:
Where is the unit test hooking the `tar` and `xz` programs?  (I can't find it in the source...is it doing it via the shell's `$PATH`?)

portante replied:
Other kinds of testing where you don't want to have to actually run the `tar` and `xz` commands, but mock out their behavior like we do in the legacy unit tests.

We have no other examples of this in the Tool Meister code where those commands are found at run time so that they can be externally overridden.

webbnh replied:
For unit testing, we shouldn't actually be running `tar`/`xz` or even external proxies for them.  We should be mocking `find_executable()` and `subprocess.Popen()`.  And, therefore, we don't care how these external functions work because we will control how the CUT sees them behave.

As for functional testing, we should be running the real things, and we shouldn't be concerned with trying to run proxies in their stead.

So, I have very little enthusiasm for setting up any kind of mid-ground, especially if the justification is that we're following the legacy testing model which we would prefer to be getting away from.

portante replied:
Let's continue this discussion in PR #2443.

portante commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -0,0 +1,257 @@
+                        self.config,
+                        self.logger,
+                    )
+                except (
+                    NotADirectoryError,
+                    FileNotFoundError,
+                    MakeResultTb.AlreadyCopied,
+                    MakeResultTb.BenchmarkRunning,
+                ) as exc:
+                    self.logger.error(str(exc))
+                    failures += 1
+                    continue
```
Not sure we can improve this much.  If an exception does blow it out of the water (the ValueError won't because we validate the host name before calling `execute()`) then the catch all exception around `execute()` will handle it.

webbnh replied:
Having the outer-most catch-all exception handler catch the error is fine (I think...), provided it can report enough information and context for us to diagnose the issue.  (Yeah, the `ValueError` is an interesting case, because we _want_ it to blow all the way back to the user, but, as you say, we explicitly forestall that by checking the controller value up front, so we don't really need to consider it here.)

We want to catch all the recoverable or result-specific failures here while letting the unrecoverable (and unexpected) ones blow by, and the question is, how do we tell the difference.  We can catch an explicit list here, but I think that's fragile; we can catch `Exception`, but that will catch the unrecoverable/unexpected ones which may or may not be OK; or, we can try to have `MakeResultTb` produce the list of exceptions to catch (presumably via a base class, but it could just export a tuple), which would do the job without violating encapsulation, and it would put the burden of maintaining the list in the appropriate place (but, it's still yucky).

portante commented on lib/pbench/agent/results.py:
```diff
@@ -122,45 +176,137 @@ def make_result_tb(self) -> str:
+                    tar_proc.wait()
+                xz_proc = None
+            else:
+                with tarball.open("w") as ofp, e_file.open("w") as efp:
+                    xz_proc = subprocess.Popen(
+                        [self.xz_path, "-T0"],
+                        cwd=str(self.target_dir),
+                        stdin=subprocess.PIPE,
+                        stdout=ofp,
+                        stderr=efp,
+                    )
+                    tar_proc = subprocess.Popen(
+                        args,
+                        cwd=str(self.result_dir.parent),
+                        stdin=None,
+                        stdout=xz_proc.stdin,
+                        stderr=efp,
+                    )
+                    tar_proc.wait()
+                    # Now that the `tar` command has exited, we close the
+                    # `stdin` of the `xz` command to shut it down.
+                    xz_proc.stdin.close()
+                    xz_proc.wait()
```
We are using `xz -T0` to parallelize the compression.  This has made a significant performance difference for folks.  This is what `pbench-move-results` does today.

webbnh replied:
OK, so the reason for invoking `xz` separately is to allow us to specify the `-T0` switch, which enables `xz` _itself_ to go multithreaded (which, I'm guessing from your comment, doesn't happen when it is invoked implicitly by `tar`), and it has nothing otherwise to do with the fact that our code runs `xz` separately from `tar`.  Fair enough.

It might be handy to mention that stuff in a code comment.

portante reviewed (commented):
> Sorry for the volume below, but it seems like these changes could use some polishing.

Agreed!  Thanks for making such an effort.

> Also, there are some weightier items:
> 
> * In general (and in the specific case cited below), do we want to document all exceptions raised by a given function?  This strikes me as a boundless problem with little payback.

Agreed, in general.  In this specific case, I found it very helpful trying to work out what and how to implement the behaviors that were required (answered inline).

> * When a function has a potentially large number of possible exceptions which all get handled in the same way, how should we construct the caller in a way which is resilient?

Not sure there is a general answer to this.  As above, it depends on how we want to implement the required behaviors.

> * Should we be using the Python2 string format specifiers, or should we stick with the Python3 ones?

Known problem.  We need to fix this ... in another PR.  =)

> * Do we want to do all the detailed handling of setting up the subprocess invocations manually, or should we just construct the appropriate shell command line and invoke it in a subprocess running a shell (which seems much simpler and easier to understand and maintain)?

It is not clear to me that it is easier to maintain.

> * Does running `tar | xz` do anything different from `tar --xz`?...is it worth supporting the two options?

`xz -T0` is different from `xz` and `tar --xz` (answered inline).

> * and a bunch of smaller stuff.

I believe I have addressed everything.

Thanks for the thorough review!

dbutenhof commented on lib/pbench/agent/results.py:
```diff
@@ -88,11 +88,13 @@ def __init__(
-                f"The benchmark is still running in {self.result_dir.name} - skipping; "
-                f"if that is not true, rmdir {self.result_dir.name}/.running, "
-                "and try again",
+                f"The benchmark is still running in {self.result_dir.name} -"
+                f" skipping; if that is not true, rmdir {running_str}, and try"
+                " again",
```
Semantically, especially with the line breaks here, it almost looks like the "if that is not true" is attached to the "skipping", and "skipping" is not subject to interpretation in this context. On one line it won't be quite so odd, but I think that we might still be better off _starting_ messages like this with the action and expanding on that with the reasoning behind the action. E.g., `"Skipping {self.result_dir.name}: the benchmark appears to be running. If that's incorrect, delete {running_str} and try again"`. (I'd rather not prescribe `rmdir` -- for example, I always use `rm -r` -- but more importantly, isn't `.running` a _file_ rather than a directory?)

dbutenhof commented on lib/pbench/agent/results.py:
```diff
@@ -226,24 +246,18 @@ def make_result_tb(self, single_threaded: bool = False) -> TarballRecord:
-                msg = f"{msg}; also encountered an error while removing failed tar ball, {tarball}: '{unlink_exc}'"
+            msg = self._unlink_tarball(
+                tarball,
+                f"Tar ball creation failed for {self.result_dir}, skipping {exc}",
```
This comes out to `skipping IOERROR` or whatever, which is not very clear. Again, I think `"Skipping {self.result_dir} because of error {exc}"` would be a lot more clear.

webbnh commented on lib/pbench/agent/results.py:
```diff
@@ -123,60 +206,143 @@ def make_result_tb(self) -> str:
 
         # The contract with the caller is to just return the full path to the
-        # created tarball.
-        return str(tarball)
+        # created tar ball.
+        return TarballRecord(tarball, tar_len, tar_md5)
```
Something tells me the "contract" has been "revised".

webbnh replied:
Since you went to the trouble of creating a named tuple, you should use the names:
```
        return TarballRecord(name=tarball, length=tar_len, md5=tar_md5)
```

webbnh commented on lib/pbench/agent/results.py:
```diff
@@ -123,60 +206,143 @@ def make_result_tb(self) -> str:
-        """CopyResultTb contructor - raises FileNotFoundError if the given
-        tar ball does not exist, and a ValueError if the given controller is
-        not a valid hostname.
+        """Contructor for object representing tar ball to be copied remotely.
```
s/Contructor/Constructor/

webbnh commented on lib/pbench/agent/results.py:
```diff
@@ -123,60 +206,143 @@ def make_result_tb(self) -> str:
         self.controller = controller
         self.tarball = Path(tarball)
         if not self.tarball.exists():
-            raise FileNotFoundError(f"Tarball '{self.tarball}' does not exist")
+            raise FileNotFoundError(f"Tar ball '{self.tarball}' does not exist")
```
It's a trivial thing, but it seems like we should do this check before assigning `self.tarball`, indeed perhaps before doing any work....

portante replied:
Yes, but why create the `Path()` object in another variable to do the test before the assignment?

webbnh commented on lib/pbench/agent/results.py:
```diff
@@ -123,60 +206,143 @@ def make_result_tb(self) -> str:
+        """Contructor for object representing tar ball to be copied remotely.
+
+        Raises
+            ValueError          if the given controller is not a valid hostname
```
I don't see where ValueError can be raised...did I miss it?  (Isn't it fun maintaining these lists?  ;-) )

webbnh commented on lib/pbench/agent/results.py:
```diff
@@ -123,60 +206,143 @@ def make_result_tb(self) -> str:
+
+        Raises
+            RuntimeError     if a connection fails to be make to the server
+            FileUploadError  if the tar balls to upload properly
```
s/balls/ball fails/

webbnh commented on lib/pbench/test/unit/agent/task/test_results_push.py:
```diff
@@ -144,38 +144,34 @@ def test_token_envar(monkeypatch, caplog, valid_config, pytestconfig):
-        assert str(result.stderr).find("Not Found") > -1
+        assert (
+            str(result.stderr).find("Not Found") > -1
+        ), f"stderr: {str(result.stderr)!r}; stdout: {str(result.stdout)!r}"
```
`result.stderr` and `result.stdout` _should_ already be strings....

portante replied:
Fixed.

webbnh commented on lib/pbench/test/unit/agent/task/test_results_move.py:
```diff
@@ -121,6 +124,49 @@ def test_move_results(monkeypatch, caplog, pytestconfig):
+        (pbrun / f"{name}.copied").unlink()
+
+        # Test --delete (default) with .running directory.
+        (pbrun / name / ".running").mkdir()
```
Just to confirm:  the `.running` marker is a _directory_ and not a file?

dbutenhof replied:
Ah; which answers one of my confusions in the prior pass. But why is it a directory? Are we putting something in it, or is that just to make it a tiny bit harder to delete accidentally?

portante replied:
Historically `.running` is always a directory.

webbnh commented on lib/pbench/test/unit/agent/task/test_results_move.py:
```diff
@@ -121,6 +124,49 @@ def test_move_results(monkeypatch, caplog, pytestconfig):
+        )
+        (pbrun / name / ".running").rmdir()
+
+        # Test --delete (default).
```
Do you want to test the `--delete`-omitted case?

portante replied:
Why wouldn't we?

webbnh replied:
> Why wouldn't we?

I took this response as "yes, we should add that", but I don't see the corresponding change to the code.

dbutenhof replied:
Neither `DELY_SWITCH` nor `DELN_SWITCH` appears in the options here which may have been Peter's intent; however with the `.running` directory, I'm not sure this is quite the test Webb envisioned as nothing happens to confirm proper defaulting. The following test case, however, seems to fit, confirming `successfully moved 1` vs [...] `copied 1`.

webbnh replied:
OK, I think maybe I am/was confused.

It now looks like `test_args()` tests the `--delete` case; and `test_results_move()` tests the `--no-delete` case, along with the default case with and without a `.running` directory present.  So, I think it's covered.

webbnh commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -175,7 +166,7 @@ def execute(self, single_threaded: bool, delete: bool = True) -> int:
 
 
-@click.command()
+@click.command(name="pbench-results-move")
```
I don't think this accomplishes anything useful.

That is, for external purposes, I think the command name is set in `setup.cfg`.  For local purposes, I expect that we're going to want the command name to be just `"move"`, because it will be a sub-command of the `"results"` subcommand.

portante replied:
Exception raised when no `name` attribute.

webbnh replied:
See [below](https://github.com/distributed-system-analysis/pbench/pull/2325#discussion_r658172074).

portante replied:
I removed the explicit `name` attribute, and the `name` becomes `main`, the name of the function.

For now, until we implement the full use of click commands with a hierarchy, etc., we want the "usage" string to say: "Usage: pbench-results-move ..." and "Usage pbench-results-push ...".

webbnh commented on lib/pbench/cli/agent/commands/results/push.py:
```diff
@@ -25,7 +25,7 @@ def execute(self) -> int:
 
 
-@click.command()
+@click.command("pbench-results-push")
```
I don't think this accomplishes anything useful.

That is, for external purposes, I think the command name is set in `setup.cfg`.  For local purposes, I expect that we're going to want the command name to be just `"push"`, because it will be a sub-command of the `"results"` subcommand.

portante replied:
Without a `name` attribute `get_default_prog_name` raises an exception.

webbnh replied:
This sounds like a symptom of another problem.  (What was the exception?)

Per the documentation, the [`name` attribute is optional](https://click.palletsprojects.com/en/8.0.x/api/#click.command), and [`get_default_prog_name`](https://click.palletsprojects.com/en/8.0.x/api/#click.testing.CliRunner.get_default_prog_name) should work if it is not supplied.

In any case, if you're going to supply an explicit name, here, I believe it should be just `"push"`, since ultimately this command will be part of the `pbench` tool in the `results` command group, and we don't want to end up with it being reported as `pbench-results-pbench-results-push`.

portante replied:
This is not something we should resolve here because the general solution for our hierarchical commands is not in place.  For now, when these commands are invoked the usage string has to say something other than `main`.

webbnh reviewed (changes requested):
Getting close.

Remaining issues:
- Consider making `tar` (and `xz`) file paths be class attributes.
- Consider reworking the `tar` invocations:
  - Let `tar` open the output file.
  - Combine the invocations into a single invocation using conditionals and the `--to-command` option.
- Address various typos and similar.

And a few nits.

ndokos commented on lib/pbench/agent/results.py:
```diff
@@ -72,23 +106,28 @@ def check_result_target_dir(dir_path: str, name: str) -> Path:
-            )
+        We then create the tar ball, verify and generate its MD5 sum value.
+
+        Returns a tuple of the Path object of the create tar ball, its length,
```
"created"

ndokos commented on lib/pbench/agent/results.py:
```diff
@@ -108,9 +147,24 @@ def make_result_tb(self) -> str:
+            # save the current controller as "controller_orig", and save the
+            # controller directory as the new controller name.
+            if md_controller:
+                mdlog.set("run", "controller_org", md_controller)
```
"controller_orig"

ndokos commented on lib/pbench/agent/results.py:
```diff
@@ -169,14 +315,17 @@ def __init__(
+                specific user.
+
+        Raises
+            RuntimeError     if a connection fails to be make to the server
```
"if a connection to the server fails to be made" perhaps?

ndokos reviewed (changes requested):
Another 24 commits to go ...

ndokos reviewed (commented):
I went through the rest of the commits and saw nothing else, so just the couple of things I found in my previous review.

webbnh commented on lib/pbench/cli/agent/commands/results/push.py:
```diff
@@ -0,0 +1,66 @@
+    try:
+        rv = ResultsPush(context).execute()
+    except Exception as exc:
+        click.echo(exc, err=True)
+        rv = 1
```
I concur...but, I'm not sure it's worth it:  we would have to figure out which exceptions should map to which error code, and then add all the appropriate conditionals, and I think it would yield very limited value.  And, I think the case that you cited is already covered (although, it also returns `1` instead of `2`, which is undesirable, but it gets dealt with before `main()` gets called.)

portante commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -0,0 +1,247 @@
+                    MakeResultTb.AlreadyCopied,
+                    MakeResultTb.BenchmarkRunning,
+                ) as exc:
+                    self.logger.error(str(exc))
```
For CLI we don't want stack traces.

portante commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -0,0 +1,247 @@
+import socket
+import tempfile
+
+import click
```
`click` is not part of the Python Standard Library.  See PEP 8.

webbnh replied:
Personally, I see no reason to distinguish between "standard libraries" and "third-party libraries" -- as far as I'm concerned, none of them are project libraries, they all come from somewhere else, and I don't see any value in having the source code (meaning _me_, as author or reviewer) have to figure which bucket to put them in.

The only argument in favor that I can see is centered on the management of dependencies...but, are we really going to look into the source code to see which block of imports a reference is in in order to generate our dependency list?  I doubt it.  (I think we'll just do a build and discover what doesn't work....)

portante commented on lib/pbench/agent/results.py:
```diff
@@ -19,7 +19,7 @@
 
 class MakeResultTb:
-    """MakeResultTb - Creates the result tar file.
+    """Interfaces for managing the creation a result tar ball.
```
Fixed.

portante commented on lib/pbench/agent/results.py:
```diff
@@ -123,60 +206,143 @@ def make_result_tb(self) -> str:
         try:
-            with tarfile.open(tarball, mode="x:xz") as tar:
-                for f in self.result_dir.rglob("*"):
-                    tar.add(os.path.realpath(f))
-        except tarfile.TarError:
-            self.logger.error(
-                "Tar ball creation failed for {}, skipping", self.result_dir
+            # Invoke tar directly for efficiency.
+            if single_threaded:
+                with tarball.open("w") as ofp, e_file.open("w") as efp:
+                    tar_proc = subprocess.Popen(
+                        args,
+                        cwd=str(self.result_dir.parent),
+                        stdin=None,
+                        stdout=ofp,
+                        stderr=efp,
+                    )
```
Not sure how `--to-command` works with generating an archive, as the `man` page seems to imply it only works with extracted files (see https://man7.org/linux/man-pages/man1/tar.1.html).

I don't see a big win to using `--file`, as that seems to cause more differences between the two cases.

I refactored the code to reduce the duplication a bit.

portante commented:
I believe I have addressed all the review comments posted so far.  I did not try to follow all the suggestions as posted.

In particular:

 * I am not a fan of class variables resolved at module load time that create bindings that need to be changed at run time for testing
 * I don't see a lot of benefit to having the `tar` command open the file such that we have a difference between the compressed vs non-compressed paths; sub-processes created do involve the parent in any I/O, and there is no code duplication

riya-17 reviewed (approved):
Looks Good

webbnh reviewed (approved):
The updates look good to me.

The `tar` subprocess creation code looks very nice, now.  I still think that we're leaving performance on the table by feeding the tarball through `stdin` instead of letting `tar` open the file itself, but that's a problem for another day, now.  (Bummer about `--to-command`...that would have been handy if it were more general.)

How we handle the paths for `tar` and `xz` doesn't have to gate this PR, but it seems odd to me.

There's one other item for consideration, below.

webbnh reviewed (changes requested):
Upon reflection, I think we need one more go-round:  I've got four or five unresolved conversations that we should settle:

- The [derivation of `AlreadyCopied` and `BenchmarkRunning`](https://github.com/distributed-system-analysis/pbench/pull/2325/files/894bac9420d2177bcafb8bd20f7699df9fbd67bc..14652f951b800f4698a02209f2bdc3353bf3e71d#r658095897)
- The Click command names ([push](https://github.com/distributed-system-analysis/pbench/pull/2325/files#r658172074) and [move](https://github.com/distributed-system-analysis/pbench/pull/2325/files#r658172538))
- A missing [test case](https://github.com/distributed-system-analysis/pbench/pull/2325#discussion_r658177835)
- And, perhaps, the [`tar` and `xz` path storage](https://github.com/distributed-system-analysis/pbench/pull/2325#discussion_r658090082).

dbutenhof commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -0,0 +1,251 @@
+                        msg = "Error uploading file"
+                    else:
+                        msg = "Unexpected error occurred copying tar ball remotely"
+                    self.logger.error(f"{msg}, '%s', %s", result_tb_name, exc)
```
Wow, that sure does look weird, combining the modern f-string syntax with the old-fashioned printf default `Logger` formatting. I'm tempted to suggest that, in this case, it would be better to defer formatting `msg` to a `%s` as part of the `Logger` formatting.

webbnh replied:
Yeah, that or do it all up-front with the f-string...but, I concur that mixing them seems less than desirable.

Take yur pick!
```
                    self.logger.error(f"{msg}, '%s', %s", result_tb_name, exc)
                    self.logger.error("%s, '%s', %s", msg, result_tb_name, exc)
                    self.logger.error("{}, '{}', {}", msg, result_tb_name, exc)
                    self.logger.error(f"{msg}, {result_tb_name!r}, {exc!s}")
```

portante replied:
We don't have that option, unfortunately.

We are in a middle ground -- while the server side has a logger that supports the use of "braces" formatting, the agent side does not have that code.  We really need to convert the agent side to share the same logger.

However, that is for another PR.

Alas, for the agent side, we can only use "%" formatting.

Hence, the layout above.

webbnh replied:
Um, OK, so the third of the four options that I offered is not viable...but why wouldn't the others (which are, I think, better than the one that you've proffered) work?

dbutenhof commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -0,0 +1,251 @@
+                        msg = "Unexpected error occurred copying tar ball remotely"
+                    self.logger.error(f"{msg}, '%s', %s", result_tb_name, exc)
+                    failures += 1
+                    # We don't know why this operation failed; regardless,
+                    # trying to copy another tar ball remotely does not have
+                    # much chance of success.
```
That's probably OK as a first pass. In reality, unless it's a connection failure or authorization failure from the server, that seems questionable. The bigger question is how often we'll get any other kind of error and whether it's worth the time and complication of decoding the possibilities.

portante replied:
Okay, I can track this in an issue going forward.

portante replied:
See issue #2429.

dbutenhof reviewed (dismissed):
I'm too lazy to go back to the conversation view to make sure I didn't say anything that has to be resolved; I think it was "cleanup" comments, so I'm approving.

webbnh commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -0,0 +1,251 @@
+    attempt to implement any backwards compatibility with that legacy
+    interface.
+
+    This command is responsible for finding all the existing pbench data
+    collection on the local host (controller), package them up as a tar ball,
+    and send to the remote pbench server.
```
How about, "This command is responsible for finding every existing pbench data collection on the local host (controller), packaging it up as a tar ball, and sending it to the remote pbench server."

dbutenhof replied:
Better, but "every existing" still at least loosely implies a collective while "packaging it up" and "sending it" are singular; "each existing" would be less ambiguous.

riya-17 commented on lib/pbench/cli/agent/commands/results/push.py:
```diff
@@ -0,0 +1,66 @@
+        super().__init__(context)
+
+    def execute(self) -> int:
+        tarball_len, tarball_md5 = md5sum(self.context.result_tb_name)
```
It seems like we will have to handle FileNotFoundError for this line as well. If the TB is non-existent then it won't be able to calculate the length and md5sum of the TB hence throwing the "No such file or directory" error.

We are checking the existence of the tarball in CopyResultTb.__init__ module.

webbnh replied:
Riya, the caller of this function has a `try` block which will handle this error...is that insufficient?

riya-17 replied:
That is but I was concerned if we really want to handle that error here and if that error will be already dealt with here, then checking for file existence in CopyResultsTb Class doesn't make any sense because that won't be raised.

webbnh replied:
I don't think that we need to handle any errors in `execute()` -- that's the elegance of reporting errors via exceptions:  they can be handled at whatever layer is appropriate, and the intermediate layers don't have to contain any code to support that.

You do have a good point that the existence check in the `CopyResultsTb` c'tor is superfluous iff we always invoke it after calling something like `md5sum()`.  However, I don't think we should be _assuming_ that that will be the case.  It's possible that the c'tor could be called in a context where the caller already has the MD5 hash from some other source, and so the fact that we have an MD5 doesn't prove that the tarball file exists.

So, in the case in question here, if the tarball file doesn't exist, then the call to `md5sum()` will raise an exception, and the call to the `CopyResultsTb` c'tor will never be made; thus, in this case, the existence check there is pointless...but, it's only two lines of code and I doubt that it's a noticeable expense, so it's probably best to retain it.

riya-17 replied:
Sure, Thank you for clarifying this :)

riya-17 reviewed (approved):
Looks Good to me

portante reviewed (commented):
I am trying to get this PR to move forward, but it has been open since May 19th, there are 193 conversations, and it seems to have entered the realm where the perfect rules the done.

I'll see how to find the rest of the conversations that are open.

portante commented:
I can't seem to find the missing test case.

webbnh commented:
> I can't seem to find the missing test case.

I'm not sure what you're referring to, here...is it one of the open conversations?

The "Conversations" pull-down in the "Files changed" tab reports 69 resolved conversations and 5 which are still open:  [1](https://github.com/distributed-system-analysis/pbench/pull/2325#discussion_r638234022), [2](https://github.com/distributed-system-analysis/pbench/pull/2325#discussion_r658095897), [3](https://github.com/distributed-system-analysis/pbench/pull/2325#discussion_r659944628), [4](https://github.com/distributed-system-analysis/pbench/pull/2325#discussion_r659974646), [5](https://github.com/distributed-system-analysis/pbench/pull/2325#discussion_r659977611).  (The pull-down gives you the dates which help with locating the outdated ones in the flow in the "Conversation" tab.)  I think you've already commented on 1, 4, and 5, so perhaps it's either 2 or 3 (but neither seems to talk about a test case).

webbnh commented on lib/pbench/cli/agent/commands/results/move.py:
```diff
@@ -22,8 +22,8 @@ class MoveResults(BaseCommand):
     This command is responsible for finding all the existing pbench data
-    collection on the local host (controller), package them up as a tar ball,
-    and send to the remote pbench server.
+    directories on the local host (controller), package each of them up as a
+    tar ball and sending it to the remote pbench server.
```
If you change things for some other reason...

s/package/packaging/

...and I prefer using the "Oxford comma" (adding a comma after "tar ball").

webbnh reviewed (changes requested):
In reference to my [previous comment](https://github.com/distributed-system-analysis/pbench/pull/2325#issuecomment-903926387), 2 is now resolved, and you've fixed 3 (although I have proposed one more tweak to it, below), and you're tracking 5 externally (so it can presumably be resoved, @dbutenhof), so there are just three conversations, left.

dbutenhof reviewed (commented):
Looks like tests 23 & 24 need gold file updates ...

webbnh reviewed (dismissed):
It all looks good to me, modulo [our quibbling](https://github.com/distributed-system-analysis/pbench/pull/2325#discussion_r658090082) over using class attributes to store paths to executables that we want to be able to mock.

I'm willing to forego that change if you would please point out to me the mechanism which causes the paths to change after the test code is loaded but before it executes the CUT.

portante reviewed (commented):
I have added issue #2443 to continue the discussion for the last remaining unresolved conversation.  I believe all other code-review feedback has been addressed, and this PR rebased on "top-of-tree" so that it is ready for final review and sign-off.

portante commented:
Not sure what is happening, but the jenkins tests failed with:

```
ERROR: Double requirement given: click (from -r /home/pbench/server/requirements.txt (line 4)) (already in click>=8.0 (from -r /home/pbench/agent/requirements.txt (line 4)), name='click')
running develop
Checking .pth file support in /var/tmp/tox/jenkins-pytests/lib/python3.8/site-packages
/usr/bin/python3 -E -c pass
TEST FAILED: /var/tmp/tox/jenkins-pytests/lib/python3.8/site-packages does NOT support .pth files
error: bad install directory or PYTHONPATH

You are attempting to install a package to a directory that is not
on PYTHONPATH and which Python does not read ".pth" files from.  The
installation directory you specified (via --install-dir, --prefix, or
the distutils default setting) was:

    /var/tmp/tox/jenkins-pytests/lib/python3.8/site-packages

and your PYTHONPATH environment variable currently contains:

    ''

Here are some of your options for correcting the problem:

* You can choose a different installation directory, i.e., one that is
  on PYTHONPATH or supports .pth files

* You can add the installation directory to the PYTHONPATH environment
  variable.  (It must then also be on PYTHONPATH whenever you run
  Python and want to use the package(s) you are installing.)

* You can set up the installation directory to support ".pth" files by
  using one of the approaches described here:

  https://setuptools.readthedocs.io/en/latest/easy_install.html#custom-installation-locations


Please make the appropriate changes for your system and try again.
```

dbutenhof commented:
> Okay, now it appears to be passing...

That's good; because we're only creating the .pth file on a full RPM install, which "can't" be happening in this environment. But I do have to wonder exactly what happened to get your poor commit into that situation.

webbnh reviewed (approved):
GTG

webbnh commented:
> Not sure what is happening, but the jenkins tests failed with:
> 
> ```
> ERROR: Double requirement given: click (from -r /home/pbench/server/requirements.txt (line 4)) (already in click>=8.0 (from -r /home/pbench/agent/requirements.txt (line 4)), name='click')
> ...
> ```

I believe the source of this was that we had "conflicting" requirements for the version of the Click package in the server and agent `requirements.txt` files.  Once Peter made the respective lines match exactly, the problem went away again.

dbutenhof commented:
I wonder if `tox` is using `.pth` files as part of its virtual environment setup, and it was thrown off kilter by the `click` conflict?
